src/dmd/cppmangle.d

```diff
@@ -602,23 +621,145 @@ private final class CppMangleVisitor : Visitor
              */
             TemplateInstance ti = d.parent.isTemplateInstance();
             assert(ti);
+            bool appendReturnType = true;
             Dsymbol p = ti.toParent();
             if (p && !p.isModule() && tf.linkage == LINK.cpp)
             {
                 buf.writeByte('N');
                 CV_qualifiers(d.type);
                 prefix_name(p);
                 if (d.isCtorDeclaration())
+                {
                     buf.writestring("C1");
+                    appendReturnType = false;
+                }
                 else if (d.isPrimaryDtor())
+                {
                     buf.writestring("D1");
+                    appendReturnType = false;
+                }
                 else
-                    source_name(ti);
+                {
+                    int firstTemplateArg = 0;
+                    bool isConvertFunc = false;
+                    string symName;
+
+                    // test for special symbols
+                    sizediff_t whichOp = -1;
+                    foreach (i, op; [Id._cast, Id.assign, Id.eq, Id.index, Id.call, Id.opUnary, Id.opBinary, Id.opOpAssign])
```

Comment:
This allocates every time, doesn't it ?

Reply:
I thought that was fixed years ago?

Reply:
nope. you can fix it though :)
() {
    enum a = [Id._cast, Id.assign, Id.eq, Id.index, Id.call, Id.opUnary, Id.opBinary, Id.opOpAssign];

Reply:
That's terrible.

Reply:
just to be sure... `immutable` will subvert TLS right? Why not `__gshared immutable array...`?

Reply:
> immutable will subvert TLS right?

Yes, it will be shared.

> Why not __gshared immutable array...

Yes, we're mostly using `__gshared` in DMD.

> () {
    static immutable array = [Id._cast, Id.assign, Id.eq, Id.index, Id.call, Id.opUnary, Id.opBinary, Id.opOpAssign];
    return array;
} ();

I don't see a need to complicate it like this, just declare it like this before the foreach loop:

```d
__gshared immutable ops = [Id._cast, Id.assign, Id.eq, Id.index, Id.call, Id.opUnary, Id.opBinary, Id.opOpAssign];
foreach (i, op; ops)
```

Reply:
Why aren't array literals in the constant data block? I really recall this being fixed forget back?

Reply:
Oh yeah, I remember why I wrote it this way now... because these `Id.[token]`'s are not static values. They are initialised at runtime, so this:
```d
__gshared immutable ops = [Id._cast, Id.assign, Id.eq, Id.index, Id.call, Id.opUnary, Id.opBinary, Id.opOpAssign];
foreach (i, op; ops)
```
Doesn't compile, because can't evaluate at compile-time.

@klickverbot says those arrays will be on stack, which feels acceptable.

Reply:
There's no reason to guess where they go. Just compile a snippet and look at the assembler. It's faster than asking, and gives a definitive answer.

Reply:
It's on the stack.

Reply:
Being on the stack makes what should be an immutable array into a performance problem, because it has to be loaded on the stack. You can make it a __gshared array and initialize it as a singleton.